features/dd-sdk-android-session-replay/src/main/kotlin/com/datadog/android/sessionreplay/internal/recorder/resources/ResourceResolver.kt

```diff
@@ -0,0 +1,328 @@
+import com.datadog.android.sessionreplay.internal.utils.DrawableUtils
+import java.util.concurrent.ExecutorService
+import java.util.concurrent.LinkedBlockingDeque
+import java.util.concurrent.ThreadPoolExecutor
+import java.util.concurrent.TimeUnit
+
+@Suppress("TooManyFunctions")
+internal class ResourceResolver private constructor(
+    private val bitmapCachesManager: BitmapCachesManager,
+    private val threadPoolExecutor: ExecutorService,
+    private val drawableUtils: DrawableUtils,
+    private val webPImageCompression: ImageCompression,
+    private val logger: InternalLogger,
+    private val md5HashGenerator: MD5HashGenerator,
+    private val recordedDataQueueHandler: DataQueueHandler,
+    private val applicationId: String,
+    private val resourceItemCreationHandler: ResourceItemCreationHandler = ResourceItemCreationHandler(
+        recordedDataQueueHandler = recordedDataQueueHandler,
+        applicationId = applicationId
+    )
+) {
+
+    // region internal
+
+    @MainThread
+    internal fun resolveResourceId(
+        applicationContext: Context,
+        displayMetrics: DisplayMetrics,
+        drawable: Drawable,
+        drawableWidth: Int,
+        drawableHeight: Int,
+        resourcesSerializerCallback: ResourcesSerializerCallback
+    ) {
+        bitmapCachesManager.registerCallbacks(applicationContext)
+
+        val resourceId = tryToGetResourceFromCache(drawable = drawable)
+
+        if (resourceId != null) {
+            // if we got here it means we saw the bitmap before,
+            // so we don't need to send the resource again
+            resourcesSerializerCallback.onReady(resourceId)
+            return
+        }
+
+        val bitmapFromDrawable = if (drawable is BitmapDrawable && shouldUseDrawableBitmap(drawable)) {
+            drawable.bitmap // cannot be null - we already checked in shouldUseDrawableBitmap
+        } else {
+        null
+        }
+
+        // do in the background
+        Runnable {
+            @Suppress("ThreadSafety") // this runs inside an executor
+            createBitmapAsync(
+                drawable = drawable,
+                drawableWidth = drawableWidth,
+                drawableHeight = drawableHeight,
+                displayMetrics = displayMetrics,
+                bitmapFromDrawable = bitmapFromDrawable,
+                resolveResourceCallback = object : ResolveResourceCallback {
+                    override fun onResolved(resourceId: String, byteArray: ByteArray) {
+                        resourceItemCreationHandler.queueItemIfNotPreviouslySeen(resourceId, byteArray)
+                        resourcesSerializerCallback.onReady(resourceId)
+                    }
+
+                    override fun onFailed() {
+                        resourcesSerializerCallback.onReady(null)
+                    }
+                }
+            )
+        }.let {
+            threadPoolExecutor.executeSafe("resolveResourceId", logger, it)
+        }
```

Comment:
minor: since lambda comes last in the `executeSafe`, we can simply do the following syntax:

```
threadPoolExecutor.executeSafe("resolveResourceId", logger) {
      @Suppress("ThreadSafety") // this runs inside an executor
      createBitmapAsync(
          drawable = drawable,
          drawableWidth = drawableWidth,
          drawableHeight = drawableHeight,
          displayMetrics = displayMetrics,
          bitmapFromDrawable = bitmapFromDrawable,
          resolveResourceCallback = object : ResolveResourceCallback {
              override fun onResolved(resourceId: String, byteArray: ByteArray) {
                  resourceItemCreationHandler.queueItemIfNotPreviouslySeen(resourceId, byteArray)
                  resourcesSerializerCallback.onReady(resourceId)
              }

              override fun onFailed() {
                  resourcesSerializerCallback.onReady(null)
              }
          }
      )
}
```